src/OpenSpaceToolkit/Astrodynamics/NumericalSolver.cpp

```diff
@@ -371,6 +389,133 @@ Array<NumericalSolver::Solution> NumericalSolver::integrateDuration(
     return integrateTime(anInitialStateVector, 0.0, aDurationArray, aSystemOfEquations);
 }
 
+NumericalSolver::Solution NumericalSolver::integrateDuration(
+    const StateVector& anInitialStateVector,
+    const Real& aDurationInSeconds,
+    const SystemOfEquationsWrapper& aSystemOfEquations,
+    const EventCondition& anEventCondition
+)
+{
+    if (stepperType_ != NumericalSolver::StepperType::RungeKuttaDopri5)
+    {
+        throw ostk::core::error::runtime::ToBeImplemented(
+            "Integrating with conditions is only supported with RungeKuttaDopri5 stepper type."
+        );
+    }
+
+    if (aDurationInSeconds.isZero())
+    {
+        return {anInitialStateVector, 0.0};
+    }
+
+    NumericalSolver::StateVector aStateVector = anInitialStateVector;
+
+    // Ensure that the time step is the correct sign
+    const double signedTimeStep = getSignedTimeStep(aDurationInSeconds);
+
+    // TBI: Adapt this to any dense stepper type
+    auto stepper = make_dense_output(absoluteTolerance_, relativeTolerance_, dense_stepper_type_5());
+
+    // initialize stepper
+    double currentTime = 0.0;
+    stepper.initialize(aStateVector, currentTime, signedTimeStep);
+
+    // do first step
+    double previousTime;
+    std::tie(previousTime, currentTime) = stepper.do_step(aSystemOfEquations);
+    observeNumericalIntegration(stepper.current_state(), stepper.current_time());
+
+    bool conditionSatisfied = false;
+    Real currentValue = Real::Undefined();
+
+    Real previousValue = anEventCondition.evaluate(stepper.current_state(), stepper.current_time());
+    NumericalSolver::StateVector currentState;
+
+    // account for integration direction
+    const auto checkTimeLimit = [&currentTime, &aDurationInSeconds]()
+    {
+        if (aDurationInSeconds > 0.0)
+        {
+            return currentTime < aDurationInSeconds;
+        }
+        return currentTime > aDurationInSeconds;
+    };
+
+    while (checkTimeLimit())
```

Comment:
Maybe we don't need this, we can maybe just check the condition, since there might always be one